src/main/java/net/minestom/server/particle/data/BlockParticle.java

```diff
@@ -0,0 +1,48 @@
+package net.minestom.server.particle.data;
+
+import net.minestom.server.instance.block.BlockState;
+import net.minestom.server.particle.ParticleType;
+import net.minestom.server.registry.Registry;
+import net.minestom.server.utils.binary.BinaryWriter;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.BiFunction;
+
+public class BlockParticle extends Particle {
+    public static final BiFunction<ParticleType<BlockParticle>, @Nullable String, BlockParticle> READER = (particle, data) -> {
```

Comment:
All particle reader seems to be fairly weak, and will likely become an issue for packet reading as they currently rely on string

Reply:
This is true, I could not find a way to have an argument change it's type dependent on the particle type. ArgumentParticle has to be divided into 2 arguments, and the second one should change it's type based on the input of the first one. The second one also shouldn't be sent to the client, because minecraft:particle has this feature already. If this is possible, it would be a better idea.